File: modules/battery-level/ios/BatteryLevelModule.swift
import ExpoModulesCore
import UIKit

public class BatteryLevelModule: Module {
  /// Block-based notification observers. `Module` is not an `NSObject`, so
  /// `#selector`-based observation is not available here.
  private var observers: [NSObjectProtocol] = []

  public func definition() -> ModuleDefinition {
    Name("BatteryLevel")

    Events("onBatteryLevelChange")

    OnCreate {
      Self.onMain { UIDevice.current.isBatteryMonitoringEnabled = true }
    }

    AsyncFunction("getBatteryLevelAsync") { () -> Float in
      return Self.currentLevel()
    }.runOnQueue(.main)

    AsyncFunction("getBatteryStateAsync") { () -> String in
      return Self.currentState()
    }.runOnQueue(.main)

    AsyncFunction("getBatteryInfoAsync") { () -> [String: Any] in
      return Self.currentInfo()
    }.runOnQueue(.main)

    // Presentational view. It receives the level as a prop rather than reading
    // the battery itself, so there is exactly one source of truth.
    View(BatteryGaugeView.self) {
      Events("onPress")

      Prop("level") { (view: BatteryGaugeView, level: Float) in
        view.level = level
      }

      Prop("charging") { (view: BatteryGaugeView, charging: Bool) in
        view.isCharging = charging
      }

      Prop("color") { (view: BatteryGaugeView, color: UIColor) in
        view.fillColor = color
      }

      Prop("trackColor") { (view: BatteryGaugeView, color: UIColor) in
        view.trackColor = color
      }

      Prop("boltColor") { (view: BatteryGaugeView, color: UIColor) in
        view.boltColor = color
      }

      OnViewDidUpdateProps { (view: BatteryGaugeView) in
        view.redraw()
      }
    }

    OnStartObserving {
      self.startObserving()
    }

    OnStopObserving {
      self.stopObserving()
    }

    OnDestroy {
      self.stopObserving()
      Self.onMain { UIDevice.current.isBatteryMonitoringEnabled = false }
    }
  }

  // MARK: - Observing

  private func startObserving() {
    guard observers.isEmpty else { return }

    Self.onMain {
      UIDevice.current.isBatteryMonitoringEnabled = true
    }

    let names: [Notification.Name] = [
      UIDevice.batteryLevelDidChangeNotification,
      UIDevice.batteryStateDidChangeNotification
    ]

    observers = names.map { name in
      NotificationCenter.default.addObserver(forName: name, object: nil, queue: .main) { [weak self] _ in
        guard let self else { return }
        self.sendEvent("onBatteryLevelChange", Self.currentInfo())
      }
    }
  }

  private func stopObserving() {
    observers.forEach { NotificationCenter.default.removeObserver($0) }
    observers.removeAll()
  }

  // MARK: - Reading the battery

  /// `UIDevice.batteryLevel` reports `-1` when monitoring is disabled or the
  /// value is unavailable (notably on the iOS Simulator). That is passed
  /// through as-is so JS can tell "unknown" apart from "empty".
  private static func currentLevel() -> Float {
    let level = UIDevice.current.batteryLevel
    return level < 0 ? -1 : level
  }

  private static func currentState() -> String {
    switch UIDevice.current.batteryState {
    case .charging:
      return "charging"
    case .full:
      return "full"
    case .unplugged:
      return "unplugged"
    default:
      return "unknown"
    }
  }

  private static func currentInfo() -> [String: Any] {
    return [
      "level": currentLevel(),
      "state": currentState()
    ]
  }

  /// Battery monitoring and `UIDevice` reads must happen on the main thread.
  private static func onMain(_ block: @escaping () -> Void) {
    if Thread.isMainThread {
      block()
    } else {
      DispatchQueue.main.async(execute: block)
    }
  }
}

File: modules/battery-level/android/src/main/java/expo/modules/batterylevel/BatteryLevelModule.kt
package expo.modules.batterylevel

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.graphics.Color
import android.os.BatteryManager
import androidx.core.content.ContextCompat
import androidx.core.os.bundleOf
import expo.modules.kotlin.exception.Exceptions
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition

class BatteryLevelModule : Module() {
  private val context: Context
    get() = appContext.reactContext ?: throw Exceptions.ReactContextLost()

  private var receiver: BroadcastReceiver? = null

  override fun definition() = ModuleDefinition {
    Name("BatteryLevel")

    Events("onBatteryLevelChange")

    AsyncFunction("getBatteryLevelAsync") {
      readLevel(currentBatteryIntent())
    }

    AsyncFunction("getBatteryStateAsync") {
      readState(currentBatteryIntent())
    }

    AsyncFunction("getBatteryInfoAsync") {
      val intent = currentBatteryIntent()
      mapOf(
        "level" to readLevel(intent),
        "state" to readState(intent)
      )
    }

    // Presentational view. It receives the level as a prop rather than reading
    // the battery itself, so there is exactly one source of truth.
    //
    // Colors arrive as hex strings and are parsed here. The built-in
    // android.graphics.Color converter requires API 26, and this project's
    // minSdk is 24.
    View(BatteryGaugeView::class) {
      Events("onPress")

      Prop("level") { view: BatteryGaugeView, level: Float ->
        view.level = level
      }

      Prop("charging") { view: BatteryGaugeView, charging: Boolean ->
        view.isCharging = charging
      }

      Prop("color") { view: BatteryGaugeView, color: String ->
        view.fillColor = parseColor(color, view.fillColor)
      }

      Prop("trackColor") { view: BatteryGaugeView, color: String ->
        view.trackColor = parseColor(color, view.trackColor)
      }

      Prop("boltColor") { view: BatteryGaugeView, color: String ->
        view.boltColor = parseColor(color, view.boltColor)
      }

      OnViewDidUpdateProps { view: BatteryGaugeView ->
        view.redraw()
      }
    }

    OnStartObserving {
      startObserving()
    }

    OnStopObserving {
      stopObserving()
    }

    OnDestroy {
      stopObserving()
    }
  }

  // MARK: - Observing

  private fun startObserving() {
    if (receiver != null) {
      return
    }

    val batteryReceiver = object : BroadcastReceiver() {
      override fun onReceive(context: Context?, intent: Intent?) {
        sendEvent(
          "onBatteryLevelChange",
          bundleOf(
            "level" to readLevel(intent),
            "state" to readState(intent)
          )
        )
      }
    }

    // ACTION_BATTERY_CHANGED is a protected system broadcast, so the receiver
    // must not be exported (required on Android 14+ / targetSdk 34+).
    ContextCompat.registerReceiver(
      context,
      batteryReceiver,
      IntentFilter(Intent.ACTION_BATTERY_CHANGED),
      ContextCompat.RECEIVER_NOT_EXPORTED
    )
    receiver = batteryReceiver
  }

  private fun stopObserving() {
    receiver?.let {
      runCatching { context.unregisterReceiver(it) }
      receiver = null
    }
  }

  // MARK: - Reading the battery

  /**
   * ACTION_BATTERY_CHANGED is sticky, so registering a `null` receiver returns
   * the last broadcast without subscribing to future ones.
   */
  private fun currentBatteryIntent(): Intent? =
    context.registerReceiver(null, IntentFilter(Intent.ACTION_BATTERY_CHANGED))

  /** Returns 0..1, or -1 when the level is unavailable — matching iOS. */
  private fun readLevel(intent: Intent?): Float {
    val level = intent?.getIntExtra(BatteryManager.EXTRA_LEVEL, -1) ?: -1
    val scale = intent?.getIntExtra(BatteryManager.EXTRA_SCALE, -1) ?: -1

    if (level < 0 || scale <= 0) {
      return -1f
    }
    return level.toFloat() / scale.toFloat()
  }

  private fun parseColor(value: String, fallback: Int): Int =
    runCatching { Color.parseColor(value) }.getOrDefault(fallback)

  private fun readState(intent: Intent?): String =
    when (intent?.getIntExtra(BatteryManager.EXTRA_STATUS, -1)) {
      BatteryManager.BATTERY_STATUS_CHARGING -> "charging"
      BatteryManager.BATTERY_STATUS_FULL -> "full"
      BatteryManager.BATTERY_STATUS_DISCHARGING,
      BatteryManager.BATTERY_STATUS_NOT_CHARGING -> "unplugged"
      else -> "unknown"
    }
}
